### PR TITLE
HDDS-8830. Add admin CLI to list open files

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

Add Ozone Admin CLI `listopenfiles` as a debug command for admins and devs. e.g.

```bash
# First, auth as Ozone admin
$ ozone admin om lof --service-id=om-service-test1 --length=3 --prefix=/volumelof/buck1
5 total open files (est.). Showing 3 open files (limit 3) under path prefix:
  /volume-lof/buck1

Client ID		Creation time	Hsync'ed	Open File Path
111726338148007937	1704808626523	No		/volume-lof/buck1/-9223372036854774527/key0
111726338151415810	1704808626578	No		/volume-lof/buck1/-9223372036854774527/key1
111726338152071171	1704808626588	No		/volume-lof/buck1/-9223372036854774527/key2

To get the next batch of open keys, run:
  ozone admin om lof -id=om-service-test1 --length=3 --prefix=/volume-lof/buck1 --start=/-9223372036854775552/-9223372036854775040/-9223372036854774527/key2/111726338152071171
```

JSON output:

```bash
$ ozone admin om lof --service-id=om-service-test1 --length=3 --prefix=/volumelof/buck1 --json
```
```json
{
  "openKeys" : [ {
    "keyInfo" : {
      "metadata" : { },
      "objectID" : -9223372036854774015,
      "updateID" : 7,
      "parentObjectID" : -9223372036854774527,
      "volumeName" : "volume-lof",
      "bucketName" : "buck1",
      "keyName" : "key0",
      "dataSize" : 4194304,
      "keyLocationVersions" : [ ... ],
      "creationTime" : 1704808722487,
      "modificationTime" : 1704808722487,
      "replicationConfig" : {
        "replicationFactor" : "THREE",
        "requiredNodes" : 3,
        "replicationType" : "RATIS"
      },
      "fileName" : "key0",
      "acls" : [ ... ],
      "path" : "-9223372036854774527/key0",
      "file" : true,
      "replicatedSize" : 12582912,
      "objectInfo" : "OMKeyInfo{volume='volume-lof', bucket='buck1', key='key0', dataSize='4194304', creationTime='1704808722487', objectID='-9223372036854774015', parentID='-9223372036854774527', replication='RATIS/THREE', fileChecksum='null}",
      "hsync" : false,
      "latestVersionLocations" : { ... },
      "updateIDset" : true
    },
    "openVersion" : 0,
    "clientId" : 111726344437039105
  }, {
    "keyInfo" : { ... },
    "openVersion" : 0,
    "clientId" : 111726344440578050
  }, {
    "keyInfo" : { ... },
    "openVersion" : 0,
    "clientId" : 111726344441233411
  } ],
  "totalOpenKeyCount" : 5,
  "hasMore" : true,
  "contToken" : "/-9223372036854775552/-9223372036854775040/-9223372036854774527/key2/111726344441233411"
}
```

- [x] Optional filter by bucket or key prefix
- [x] Pagination support
- [x] JSON output
- [x] Documentation

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8830

## How was this patch tested?

- [x] Integration test addition.
- [x] Unit test addition.

### CODE REVIEW

**Review by jojochuang — COMMENTED** [review]
Looks like this would list both visible and invisible (keys that are written but no hsync() is called at it) keys. If that's the intend it would be good to distinguish them in the output.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 1]
```diff
@@ -3186,6 +3192,132 @@ public ServiceInfoEx getServiceInfo() throws IOException {
...
+    }
+
+    final String dbOpenKeyPrefix;
+    final Table<String, OmKeyInfo> openKeyTable;
+    if (path == null || path.isEmpty() || path.equals(OM_KEY_PREFIX)) {
+      // path is root
+      dbOpenKeyPrefix = "";
+      // default to FSO. TODO: Add client option to pass OBS/LEGACY?
+      openKeyTable =
+          metadataManager.getOpenKeyTable(BucketLayout.FILE_SYSTEM_OPTIMIZED);
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:3220]: I think it depends on whether or not we want to export only keys visible in the namespace. If so, then this semantics is only useful for FSO.  If not, there could be use cases where it makes sense for legacy/OBS

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+      names = {"-n"},
+      description = "Numerical limit of open files/keys to return.",
+      defaultValue = "1000",
+      hidden = false
+  )
+  private long count;
+
+  @CommandLine.Option(
+      names = {"-t", "--token"},  // TODO: Conform to Ozone CLI convention
+      description = "Previous/last file/key path as the continuation token.",
```
Comment (jojochuang) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java:88]: Add an example?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java [hunk 3]
```diff
@@ -549,6 +551,44 @@ public void testOzoneAdminCmdList() throws UnsupportedEncodingException {
...
+    try (FSDataOutputStream out = fs.create(new Path(key1))) {
+      out.write(1);
+      // Wait for flush
+//      cluster.getOzoneManager().awaitDoubleBufferFlush();
+
+      // Hold the file open, run listopenfiles
+      String[] args = new String[] {"om", "listopenfiles", "-id", omServiceId,
+          "-p", "/volumelof/buck1"};
+      // TODO: Possible to make this work under volume level as well?
+      execute(ozoneAdminShell, args);
```
Comment (jojochuang) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java:581]: please run listopenfiles again after out.hsync()

**Comment by smengcl** [comment]
> Looks like this would list both visible and invisible (keys that are written but no hsync() is called at it) keys. If that's the intend it would be good to distinguish them in the output.

That makes sense. Added in the latest revision. Thanks for taking a look at the draft.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 4]
```diff
@@ -3186,6 +3192,132 @@ public ServiceInfoEx getServiceInfo() throws IOException {
     return serviceInfo.provide();
   }
 
+  @Override
+  public ListOpenFilesResult listOpenFiles(String path,
+                                           long maxKeys,
+                                           String contToken)
+      throws IOException {
+
+//    metrics.incListOpenFiles();  // TODO: Do we want a counter for this?
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:3201]: Donee

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java]: Users of this command won't have to remember the usage of this option's usage ideally, as I have added `getCmdForNextBatch()` to prompt the command to fetch the next batch. User could just copy, paste and run.

Markdown doc to be added later.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 5]
```diff
@@ -3186,6 +3192,132 @@ public ServiceInfoEx getServiceInfo() throws IOException {
...
+          "Only Ozone admins are allowed to list open files.",
+          PERMISSION_DENIED);
+      AUDIT.logWriteFailure(buildAuditMessageForFailure(
+          OMAction.LIST_OPEN_FILES, new LinkedHashMap<>(), omEx));
+      throw omEx;
+    }
+
+    final String dbOpenKeyPrefix;
+    final Table<String, OmKeyInfo> openKeyTable;
+    if (path == null || path.isEmpty() || path.equals(OM_KEY_PREFIX)) {
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:3219]: IMO it is more flexible (in terms of not having to break protobuf msg) to parse it on the server side.

**Comment by jojochuang** [comment]
You can check out OmMetadataManagerImpl.getExpiredOpenKeys()   https://github.com/apache/ozone/blob/f4a3b0e64c0cb3f2debaded68dcc0b7573602837/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L1808
to see how we tell a key has hsync or not. You shouldn't need to walk through the keyTable.

**Comment by errose28** [comment]
Just a few comments on the CLI portion of this change:

- I think we should follow standard kebab-case conventions for commands going forwards.
-  We should avoid combining multiple words in one command since it makes it less flexible for future use. Instead of `ozone admin om listopenfiles` we could do something like `ozone admin om open-files ls` in case we need to provide more open file operations in the future. 
- I'm not sure whether we want to use `open-files` or `open-keys` as the term here. Even though this is being implemented for hsync which is a filesystem operation, the command is under `ozone admin` and therefore outside of ofs. In this area we usually refer to objects as `keys` like `ozone admin key list`.
- Related to the above, can you clarify the scope of this command? Is it operating on only open files, open keys as well? Does it work for all bucket layouts? Do you pass it a bucket name to limit the scope?
- Can you provide an example of expected CLI parameters and output in the PR description? This makes it easier for others to quickly give feedback on the UI portion without reviewing implementation details.
- Since the output will likely be verbose, I think it makes sense to only have this command print json, similar to `ozone sh key list`. In the current draft there is a --json flag that defaults to false.

**Comment by smengcl** [comment]
Thanks @errose28 for the comment.

> * I think we should follow standard kebab-case conventions for commands going forwards.
> * We should avoid combining multiple words in one command since it makes it less flexible for future use. Instead of `ozone admin om listopenfiles` we could do something like `ozone admin om open-files ls` in case we need to provide more open file operations in the future.

+1 this. But in this way we'd also need to rename all other commands (e.g. `getserviceroles`, `cancelprepare`). I'd like @jojochuang and @adoroszlai 's opinions as well.

> * I'm not sure whether we want to use `open-files` or `open-keys` as the term here. Even though this is being implemented for hsync which is a filesystem operation, the command is under `ozone admin` and therefore outside of ofs. In this area we usually refer to objects as `keys` like `ozone admin key list`.

This is basically a command that enumerates `OpenKeyTable` (or `OpenFileTable`). It is not just limited to hsync'ed keys but all open keys. I should remove `[hsync]` from the jira title.

I do like the idea of `ozone admin key list --open`.

> * Related to the above, can you clarify the scope of this command? Is it operating on only open files, open keys as well? Does it work for all bucket layouts? Do you pass it a bucket name to limit the scope?

Yup it is supposed to work with FSO/OBS/LEGACY. Path to a bucket (or just a key prefix) can be optionally passed to limit the scope. But without it (or when passing root as the path) it just iterates over two `OpenKey(File)Table`s.

> * Can you provide an example of expected CLI parameters and output in the PR description? This makes it easier for others to quickly give feedback on the UI portion without reviewing implementation details.

I planned to. But it is not finalized yet so I will do that later. Plus I will add it to the markdown doc as well.

> * Since the output will likely be verbose, I think it makes sense to only have this command print json, similar to `ozone sh key list`. In the current draft there is a --json flag that defaults to false.

I think we need both.

**Comment by smengcl** [comment]
> You can check out OmMetadataManagerImpl.getExpiredOpenKeys()
> 
> https://github.com/apache/ozone/blob/f4a3b0e64c0cb3f2debaded68dcc0b7573602837/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L1808
> 
> 
> to see how we tell a key has hsync or not. You shouldn't need to walk through the keyTable.

As I can see it is using the same approach as mine to check whether a key is hsync'ed or not. AFAIK the metadata tag is the only way to do it at the moment. Plus `getExpiredOpenKeys` looks to be iterating over the entire `KeyTable`/`FileTable`.

Again, the problem is that `OpenKeyTable` doen't have this metadata tag at all. If we can add that it will be much more efficient (both for `getExpiredOpenKeys` and this PR).

**Comment by smengcl** [comment]
Filed HDDS-10077 for hsync metadata tag improvement.

**Comment by smengcl** [comment]
Looks like branch HDDS-7593 needs a master branch merge before I can cleanly base this PR on to that one.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOmMetadataManager.java [hunk 6]
```diff
@@ -558,6 +561,128 @@ public void testListKeysWithFewDeleteEntriesInCache() throws Exception {
...
+    testListOpenFiles(BucketLayout.LEGACY);
+  }
+
+  /**
+   * Tests inner impl of listOpenFiles with different bucket types with and
+   * without pagination. NOTE: This UT does NOT test hsync in this since hsync
+   * status check is done purely on the client side.
+   * @param bucketLayout BucketLayout
+   */
+  public void testListOpenFiles(BucketLayout bucketLayout) throws Exception {
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOmMetadataManager.java:587]: Let's use `@ParameterizedTest` for the same.

```suggestion
  /**
   * Tests inner impl of listOpenFiles with different bucket types with and
   * without pagination. NOTE: This UT does NOT test hsync in this since hsync
   * status check is done purely on the client side.
   */
  @ParameterizedTest
  @EnumSource
  void testListOpenFiles(BucketLayout bucketLayout) throws Exception {
```

(additional import needed for `EnumSource`)

**Review by adoroszlai — COMMENTED** [review]
Thanks @smengcl for the patch.

**Comment by errose28** [comment]
Right now this is returning object IDs in the path of the keys, which were previously only internal to OM. What is the expected use of this CLI? If it is to do some sort of file close/reclamation based on key name the current paths returned will not help. Alternatively, providing the actual full paths may be problematic because keys can remain open after their parent directories or buckets are deleted, and only the final commit will fail.

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 7]
```diff
@@ -1523,6 +1529,24 @@ message CancelPrepareResponse {
 
 }
 
+message ListOpenFilesRequest {
+  optional string path = 1;
```
Comment (jojochuang) [inline hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:1533]: is it really optional ...? As far as I can tell the OM code expects these three parameter all the time.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMetrics.java [hunk 8]
```diff
@@ -428,10 +430,18 @@ public void incNumListS3Buckets() {
   }
 
   public void incNumListS3BucketsFails() {
-    numBucketOps.incr();
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMetrics.java:431]: Unrelated change?

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto (hunk 7 above)
Comment (smengcl) [reply hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:1533]: It is for proto3 migration at some point. Same idea as this:

https://github.com/apache/ozone/pull/3824#discussion_r1011376445

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMetrics.java (hunk 8 above)
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMetrics.java:431]: Correct. Haven't bothered to file a new jira to do this one-line fix.

`Fails` metric isn't supposed to increase `Ops`.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+  @CommandLine.Option(
+      names = {"-id", "--service-id"},
+      description = "Ozone Manager Service ID",
+      required = false
+  )
+  private String omServiceId;
+
+  @CommandLine.Option(
+      names = {"-host", "--service-host"},
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java:57]: Yup I did realized the naming convention is not conformant when I copied this chunk from `UpdateRangerSubcommand`. Thought we might one day do a batch correction on those?

Let me remove the short hand here at least.

[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/ListOpenFilesSubCommand.java:57]: Thanks.  Yes, we should fix them sometime.  The problem is that existing commands/options need deprecation etc.  That's why I try to prevent any new commands introduce options that should be deprecated right away...

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOmMetadataManager.java [hunk 10]
```diff
@@ -558,6 +561,128 @@ public void testListKeysWithFewDeleteEntriesInCache() throws Exception {
 
   }
 
+  @Test
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOmMetadataManager.java:564]: These three test cases can be parameterized using JUnit5.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 11]
```diff
@@ -3187,14 +3192,96 @@ public ServiceInfoEx getServiceInfo() throws IOException {
...
+      volumeName = tokenizer.nextToken();
+      bucketName = tokenizer.nextToken();
+
+      OmBucketInfo bucketInfo;
+      try {
+        // as expected, getBucketInfo throws if volume or bucket does not exist
+        bucketInfo = getBucketInfo(volumeName, bucketName);
+      } catch (Exception ex) {
+        metrics.incNumListOpenFilesFails();
+        throw ex;
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:3237]: consider wrapping it in an OMException.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java [hunk 12]
```diff
@@ -549,6 +555,134 @@ public void testOzoneAdminCmdList() throws UnsupportedEncodingException {
...
+      cmdRes = getStdOut();
+
+      // Verify that only one key is hsync'ed
+      assertTrue(cmdRes.contains("\tYes\t"), "One key should be hsync'ed");
+      assertTrue(cmdRes.contains("\tNo\t"), "One key should not be hsync'ed");
+    } finally {
+      // Cleanup
+      for (int i = 0; i < numKeys; i++) {
+        streams[i].close();
+      }
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java:672]: nice!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 11 above)
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:3237]: done

**Review by jojochuang — APPROVED** [review]
LGTM. Merging this one so we can begin testing in actual environments.

**Comment by smengcl** [comment]
Thanks @jojochuang for merging this to the HDDS-7593 feature branch. And thanks @adoroszlai and @errose28 for the reviews and comments.

Will post a PR for HDDS-10077 that depends on this one.
